clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -358,14 +358,17 @@ public void wakeup() {
     @Override
     public void close() {
         List<String> connections = new ArrayList<>(channels.keySet());
-        for (String id : connections)
-            close(id);
         try {
-            this.nioSelector.close();
-        } catch (IOException | SecurityException e) {
-            log.error("Exception closing nioSelector:", e);
+            for (String id : connections)
+                close(id);
+            try {
+                this.nioSelector.close();
+            } catch (IOException | SecurityException e) {
+                log.error("Exception closing nioSelector:", e);
+            }
+        } finally {
+            sensors.close();
```

Comment:
Since we have several things that all need to be closed, perhaps we could use `ClientUtils.closeQuietly`? Maybe something like this
```java
try {
  for (String id : connections)
    close(id);
} finally {
    AtomicReference<Throwable> firstException = new AtomicReference<>();
    closeQuietly(nioSelector, firstException);
    closeQuietly(sensors, firstException);
    closeQuietly(channelBuilder, firstException)
    if (firstException.get() != null)
      throw firstException.get()
}
```